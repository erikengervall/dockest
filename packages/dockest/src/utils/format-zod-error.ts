import { ZodError } from 'zod';
import { fromZodError } from 'zod-validation-error';

/**
 * Format a ZodError into a string with a newline per issue
 *
 * Completely missing fields in an object will be output as `Required at "path.to.field"`
 *
 * There's a hard limit of 100 issues in the message, so if you have more than 100 issues, you'll get a truncated message.
 */
export function formatZodError(
  zodError: ZodError,
  /**
   * The Zod model's name, E.g. "Person".
   */
  modelName: string,
) {
  const { message, name } = fromZodError(zodError, {
    issueSeparator: '\n',
    prefix: '',
    prefixSeparator: '',
    unionSeparator: '\n',
    maxIssuesInMessage: 100,
  });

  return `[${name} for "${modelName}"]
${message}`;
}
