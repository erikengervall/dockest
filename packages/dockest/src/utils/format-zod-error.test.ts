import { z } from 'zod'

import { formatZodError } from './format-zod-error'

describe('formatZodError', () => {
  const Person = z.object({
    name: z.string(),
    age: z.number(),
    nested: z.object({
      field: z.string(),
    }),
  })
  const Options = z.union([z.literal('option-1'), z.literal('option-2')])

  it('should format a ZodError for a union', () => {
    const invalidOptions = Options.safeParse('not-an-option')
    if (invalidOptions.success) {
      fail("Shouldn't reach this point")
    }

    const result = formatZodError(invalidOptions.error, 'Options')

    expect(result).toMatchInlineSnapshot(`
      "[ZodValidationError for "Options"]
      Invalid literal value, expected "option-1"
      Invalid literal value, expected "option-2""
    `)
  })

  it('should format a ZodError with an identity', () => {
    const invalidPerson = Person.safeParse({ name: 1 })
    if (invalidPerson.success) {
      fail("Shouldn't reach this point")
    }

    const result = formatZodError(invalidPerson.error, 'Person')

    expect(result).toMatchInlineSnapshot(`
      "[ZodValidationError for "Person"]
      Expected string, received number at "name"
      Required at "age"
      Required at "nested""
    `)
  })

  it('should format a ZodError without an identity', () => {
    const invalidPerson = Person.safeParse({ name: 1 })
    if (invalidPerson.success) {
      fail("Shouldn't reach this point")
    }

    const result = formatZodError(invalidPerson.error, 'Person')

    expect(result).toMatchInlineSnapshot(`
      "[ZodValidationError for "Person"]
      Expected string, received number at "name"
      Required at "age"
      Required at "nested""
    `)
  })

  it('should format a ZodError with several errors', () => {
    const Person = z.object({
      name: z.string().min(5),
      age: z.number().min(33),
    })
    const invalidPerson = Person.safeParse({
      name: 'Rob',
      age: 32,
    })
    if (invalidPerson.success) {
      fail("Shouldn't reach this point")
    }

    const result = formatZodError(invalidPerson.error, 'Person')

    expect(result).toMatchInlineSnapshot(`
      "[ZodValidationError for "Person"]
      String must contain at least 5 character(s) at "name"
      Number must be greater than or equal to 33 at "age""
    `)
  })

  it('should truncate if more than 100 issues', () => {
    const Person = z.array(
      z.object({
        name: z.string().min(5),
        age: z.number().min(33),
      }),
    )
    const invalidPerson = Person.safeParse(
      Array.from({ length: 300 }).map(() => ({
        name: 'Rob',
        age: 32,
      })),
    )
    if (invalidPerson.success) {
      fail("Shouldn't reach this point")
    }

    const result = formatZodError(invalidPerson.error, 'Person')

    expect(result.split('\n').length).toMatchInlineSnapshot(`101`)
    expect(result).toMatchSnapshot()
  })
})
