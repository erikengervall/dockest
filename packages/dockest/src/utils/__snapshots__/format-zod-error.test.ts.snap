// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatZodError should truncate if more than 100 issues 2`] = `
"[ZodValidationError for "Person"]
String must contain at least 5 character(s) at "[0].name"
Number must be greater than or equal to 33 at "[0].age"
String must contain at least 5 character(s) at "[1].name"
Number must be greater than or equal to 33 at "[1].age"
String must contain at least 5 character(s) at "[2].name"
Number must be greater than or equal to 33 at "[2].age"
String must contain at least 5 character(s) at "[3].name"
Number must be greater than or equal to 33 at "[3].age"
String must contain at least 5 character(s) at "[4].name"
Number must be greater than or equal to 33 at "[4].age"
String must contain at least 5 character(s) at "[5].name"
Number must be greater than or equal to 33 at "[5].age"
String must contain at least 5 character(s) at "[6].name"
Number must be greater than or equal to 33 at "[6].age"
String must contain at least 5 character(s) at "[7].name"
Number must be greater than or equal to 33 at "[7].age"
String must contain at least 5 character(s) at "[8].name"
Number must be greater than or equal to 33 at "[8].age"
String must contain at least 5 character(s) at "[9].name"
Number must be greater than or equal to 33 at "[9].age"
String must contain at least 5 character(s) at "[10].name"
Number must be greater than or equal to 33 at "[10].age"
String must contain at least 5 character(s) at "[11].name"
Number must be greater than or equal to 33 at "[11].age"
String must contain at least 5 character(s) at "[12].name"
Number must be greater than or equal to 33 at "[12].age"
String must contain at least 5 character(s) at "[13].name"
Number must be greater than or equal to 33 at "[13].age"
String must contain at least 5 character(s) at "[14].name"
Number must be greater than or equal to 33 at "[14].age"
String must contain at least 5 character(s) at "[15].name"
Number must be greater than or equal to 33 at "[15].age"
String must contain at least 5 character(s) at "[16].name"
Number must be greater than or equal to 33 at "[16].age"
String must contain at least 5 character(s) at "[17].name"
Number must be greater than or equal to 33 at "[17].age"
String must contain at least 5 character(s) at "[18].name"
Number must be greater than or equal to 33 at "[18].age"
String must contain at least 5 character(s) at "[19].name"
Number must be greater than or equal to 33 at "[19].age"
String must contain at least 5 character(s) at "[20].name"
Number must be greater than or equal to 33 at "[20].age"
String must contain at least 5 character(s) at "[21].name"
Number must be greater than or equal to 33 at "[21].age"
String must contain at least 5 character(s) at "[22].name"
Number must be greater than or equal to 33 at "[22].age"
String must contain at least 5 character(s) at "[23].name"
Number must be greater than or equal to 33 at "[23].age"
String must contain at least 5 character(s) at "[24].name"
Number must be greater than or equal to 33 at "[24].age"
String must contain at least 5 character(s) at "[25].name"
Number must be greater than or equal to 33 at "[25].age"
String must contain at least 5 character(s) at "[26].name"
Number must be greater than or equal to 33 at "[26].age"
String must contain at least 5 character(s) at "[27].name"
Number must be greater than or equal to 33 at "[27].age"
String must contain at least 5 character(s) at "[28].name"
Number must be greater than or equal to 33 at "[28].age"
String must contain at least 5 character(s) at "[29].name"
Number must be greater than or equal to 33 at "[29].age"
String must contain at least 5 character(s) at "[30].name"
Number must be greater than or equal to 33 at "[30].age"
String must contain at least 5 character(s) at "[31].name"
Number must be greater than or equal to 33 at "[31].age"
String must contain at least 5 character(s) at "[32].name"
Number must be greater than or equal to 33 at "[32].age"
String must contain at least 5 character(s) at "[33].name"
Number must be greater than or equal to 33 at "[33].age"
String must contain at least 5 character(s) at "[34].name"
Number must be greater than or equal to 33 at "[34].age"
String must contain at least 5 character(s) at "[35].name"
Number must be greater than or equal to 33 at "[35].age"
String must contain at least 5 character(s) at "[36].name"
Number must be greater than or equal to 33 at "[36].age"
String must contain at least 5 character(s) at "[37].name"
Number must be greater than or equal to 33 at "[37].age"
String must contain at least 5 character(s) at "[38].name"
Number must be greater than or equal to 33 at "[38].age"
String must contain at least 5 character(s) at "[39].name"
Number must be greater than or equal to 33 at "[39].age"
String must contain at least 5 character(s) at "[40].name"
Number must be greater than or equal to 33 at "[40].age"
String must contain at least 5 character(s) at "[41].name"
Number must be greater than or equal to 33 at "[41].age"
String must contain at least 5 character(s) at "[42].name"
Number must be greater than or equal to 33 at "[42].age"
String must contain at least 5 character(s) at "[43].name"
Number must be greater than or equal to 33 at "[43].age"
String must contain at least 5 character(s) at "[44].name"
Number must be greater than or equal to 33 at "[44].age"
String must contain at least 5 character(s) at "[45].name"
Number must be greater than or equal to 33 at "[45].age"
String must contain at least 5 character(s) at "[46].name"
Number must be greater than or equal to 33 at "[46].age"
String must contain at least 5 character(s) at "[47].name"
Number must be greater than or equal to 33 at "[47].age"
String must contain at least 5 character(s) at "[48].name"
Number must be greater than or equal to 33 at "[48].age"
String must contain at least 5 character(s) at "[49].name"
Number must be greater than or equal to 33 at "[49].age""
`;
